### PR TITLE
fix(glsl): reflect promoted subgraph widget edits in live preview and histogram

### PR DESCRIPTION
**Pull request by jtydhr88**
Under ADR 0009 link-only promotion, a promoted proxy widget's live value
lives at the host widget key while its node output (histogram) is keyed
by the interior source node's locator. Two consumers read the wrong key:

- GLSL live preview read the interior widget key, so editing a promoted curve/range proxy never re-rendered. extractUniformSources now resolves each promoted source to its host widget key via findHostInputForPromotion and curveValues/collectValues read that first.
- Curve/range histogram read the host node locator. getWidgetNodeLocatorId now resolves promoted widgets to their source node's locator via executionIdToNodeLocatorId.

## Screenshots
before

https://github.com/user-attachments/assets/095dfb68-143e-45f1-91b4-63be2400b46d

after

https://github.com/user-attachments/assets/e2a02d89-d001-420d-b472-8c7b0a5b9e45

### CODE REVIEW

File: src/renderer/glsl/useGLSLUniforms.ts [hunk 1]
```diff
@@ -95,6 +100,10 @@ export function extractUniformSources(
     const source: UniformSource = {
       nodeId: sourceNode.id,
       widgetName: widget.name,
+      hostWidgetId: subgraphNode
+        ? findHostInputForPromotion(subgraphNode, sourceNode.id, widget.name)
```
Comment (christian-byrne) [inline src/renderer/glsl/useGLSLUniforms.ts]: Resolution is correct: `findHostInputForPromotion(subgraphNode, sourceNode.id, widget.name)?.widgetId` maps the interior source widget to its promoted host proxy key (ADR-0009 — the proxy's live value lives at the host key). Nicely guarded: when `subgraphNode` is undefined (the non-subgraph/direct-node path) `hostWidgetId` stays undefined and the original interior path is preserved.

File: src/renderer/glsl/useGLSLUniforms.ts [hunk 2]
```diff
@@ -133,12 +142,14 @@ export function useGLSLUniforms(
...
-      return subgraphSources.map(({ nodeId: nId, widgetName, directValue }) => {
-        const widget = widgetValueStore.getWidget(
-          widgetId(gId, nId, widgetName)
-        )
-        return coerce(widget?.value ?? directValue() ?? defaultValue)
-      })
+      return subgraphSources.map(
+        ({ nodeId: nId, widgetName, hostWidgetId, directValue }) => {
+          const widget = widgetValueStore.getWidget(
+            hostWidgetId ?? widgetId(gId, nId, widgetName)
```
Comment (christian-byrne) [inline src/renderer/glsl/useGLSLUniforms.ts:149]: This is the core seam of the fix — preferring `hostWidgetId` over the interior `widgetId(gId, nId, widgetName)` is what makes a promoted proxy edit reach the renderer, and the fallback keeps non-promoted nodes on the old path. Note this read (and its twin in `curveValues` at L233) is **not** pinned by any test: `useGLSLUniforms.test.ts` only asserts `extractUniformSources` *populates* `hostWidgetId`. A future revert of just these two `??` reads would keep those tests green while re-breaking the preview. I authored a red/green read-side guard that fails if either is dropped — offering it in a top-level comment.

File: src/renderer/extensions/vueNodes/composables/useProcessedWidgets.ts [hunk 3]
```diff
@@ -185,8 +188,18 @@ function getProcessedNodeExecutionId(
 
 function getWidgetNodeLocatorId(
   nodeData: VueNodeData,
-  bareWidgetId: NodeId | null
+  bareWidgetId: NodeId | null,
+  sourceExecutionId: NodeExecutionId | undefined,
+  rootGraph: LGraph | null
 ): NodeLocatorId | undefined {
+  if (sourceExecutionId && rootGraph) {
```
Comment (christian-byrne) [inline src/renderer/extensions/vueNodes/composables/useProcessedWidgets.ts:195]: This is the opposite-direction half of the same root cause: the locator was reading the *host* node, so redirecting via `executionIdToNodeLocatorId(rootGraph, sourceExecutionId)` to the *source* interior node is the right fix for the histogram. Early-return only when a source locator resolves, otherwise it falls through to `bareWidgetId` — non-promoted widgets are unaffected. LGTM.

File: src/renderer/glsl/useGLSLUniforms.test.ts [hunk 4]
```diff
@@ -1,14 +1,23 @@
...
 import type { LGraphNode } from '@/lib/litegraph/src/LGraphNode'
 import type { Subgraph } from '@/lib/litegraph/src/subgraph/Subgraph'
+import type { SubgraphNode } from '@/lib/litegraph/src/subgraph/SubgraphNode'
 
 import {
   extractUniformSources,
   toNumber
 } from '@/renderer/glsl/useGLSLUniforms'
 
+const hostLookup = vi.hoisted(
```
Comment (christian-byrne) [inline src/renderer/glsl/useGLSLUniforms.test.ts]: Echoing CodeRabbit's Major (test-hygiene, cosmetic — not about the fix): `hostLookup` is a mutable module-level wrapper reset by hand at L143. If an assertion throws before that reset runs, the stub leaks into the next test. Prefer a `vi.hoisted()` `vi.fn()` and set expectations per-test with `mockReturnValueOnce` / `mockImplementationOnce` so cleanup is automatic and there's no shared mutable state. Same pattern applies to the `graphTraversalUtil` mock in `useProcessedWidgets.test.ts` (L30-34). Non-blocking, but it's the one thing holding CodeRabbit's CHANGES_REQUESTED.

**Review by christian-byrne — COMMENTED** [review]
## Review — TRUE root-cause fix at the ADR-0009 promotion-resolution layer

Confirmed this is a genuine root-cause fix, not a band-aid. It resolves both halves of the same underlying bug: under ADR-0009 link-only promotion the promoted proxy's live value lives at the **host** widget key (`graphId:hostNodeId:name`) while the interior source node stays the discovery/execution anchor. Two consumers read the wrong key, in opposite directions:

1. **GLSL live preview** (`useGLSLUniforms`) discovered sources from the interior node and read `widgetId(gId, interiorId, name)`, so a promoted curve/range proxy edit never reached the renderer. Fixed by threading `subgraphNode` into `extractUniformSources`, resolving `hostWidgetId` via `findHostInputForPromotion(...)?.widgetId`, and preferring it at the read sites (`hostWidgetId ?? widgetId(...)`).
2. **Curve/range histogram** (`useProcessedWidgets.getWidgetNodeLocatorId`) read the *host* node locator. Fixed by redirecting to the source interior node's locator via `executionIdToNodeLocatorId(rootGraph, sourceExecutionId)`.

Both changes are additive and guarded by `??` / early-return fallbacks, so non-promoted and non-subgraph nodes take the original path unchanged — regression risk is low. No file overlap with #13851 (`useStringWidget.ts`) or #13874 (new test file only). CI is fully green.

### Correctness / adversarial notes
- **Graceful degradation:** if `findHostInputForPromotion` returns undefined (e.g. host proxy not yet registered), `hostWidgetId` stays undefined and the read falls back to the interior key — i.e. the pre-fix behavior, never worse.
- **Value fallback chain** `widget?.value ?? directValue() ?? defaultValue` is preserved, so a host store entry with a nullish value still falls through to the interior `directValue()`. Fine, but worth being aware of.
- **Coverage gap (the one substantive gap):** the added unit tests only pin that `extractUniformSources` *populates* `hostWidgetId` and that the locator resolves to the source node. Nothing pins the **read** side — that `collectValues`/`curveValues` actually *prefer* `hostWidgetId`. A future revert of just the two `hostWidgetId ??` reads would keep every existing test green while silently re-breaking the preview. I've authored a red/green guard that closes exactly this seam (see comment below) and verified it goes red if either `??` is dropped.

### Merge gate
- CodeRabbit `CHANGES_REQUESTED` is a single **Major test-hygiene** finding (cosmetic, not about the fix) — see inline. Addressing it should clear the block.
- Pre-merge "End-To-End Regression Coverage For Fixes" is inconclusive; an e2e editing a promoted curve/range proxy and asserting the preview re-renders would satisfy it (existing `glslPreview.spec.ts` only covers direct edits).

Labels `needs-backport` + `core/1.47` + `cloud/1.47` are present, so this auto-backports on merge — the 1.47 RC won't carry Terry's fix until this lands.

**Comment by christian-byrne** [comment]
@jtydhr88 I pushed one commit to this branch (`3957fc3`) adding `src/renderer/glsl/useGLSLUniforms.promotedRead.test.ts` — a red/green regression guard that closes the **read-side** coverage gap I flagged in my review.

Your existing tests pin that `extractUniformSources` *populates* `hostWidgetId` and that the locator resolves to the source node — the plumbing. But nothing pins that `collectValues`/`curveValues` actually **prefer** `hostWidgetId` over the interior key. This test drives the real `useGLSLUniforms` composable + real `widgetValueStore` (no store mock): it seeds an *edited* value at the host key and a *stale* value at the interior key, then asserts the host value is returned.

Verified locally against this branch head:
- **GREEN** as-is (`Tests 2 passed`).
- **RED** if either `hostWidgetId ?? widgetId(gId, nId, widgetName)` read is reverted to the interior key (`expected [7] to deeply equal [42]`).

It's lint + oxfmt clean. Complementary to your plumbing tests — happy to fold it into `useGLSLUniforms.test.ts` instead if you'd prefer, but it's kept standalone because it needs real pinia/store and no `promotionUtils` mock. Feel free to drop the commit if you'd rather own the read-side coverage yourself.

**Comment by christian-byrne** [comment]
Picking this up on Terry's behalf (he's out of town) so it can land today.

What was failing: the pushed `useGLSLUniforms.promotedRead.test.ts` imported `NodeId` from `@/lib/litegraph/src/LGraphNode` (not exported) and `UUID` from `@/lib/litegraph/src/utils/uuid` (unresolvable), so `vue-tsc` failed (TS2459/TS2307) in `setup` and cascaded to every downstream job; knip also flagged the unresolved import.

Fixes pushed:
- Corrected the imports (`NodeId` from `@/types/nodeId`, `UUID` from `@/utils/uuid`).
- Addressed CodeRabbit's test-hygiene item: replaced the mutable `hostLookup` wrapper / always-on module mock with `vi.hoisted()` `vi.fn()` mocks and per-test `mockImplementationOnce`/`mockReturnValueOnce` (no global mutable mock state) in `useGLSLUniforms.test.ts` and `useProcessedWidgets.test.ts`.
- Added an e2e (`browser_tests/tests/vueNodes/glslPreview.spec.ts`) that promotes an interior float onto the subgraph host node, edits the promoted proxy on the host, and asserts the live GLSL preview re-renders from the host widget value — end-to-end guard for the read-side fix.
- Merged current `main` in (was 12 behind; clean merge).

Unit coverage is red/green verified: `promotedRead` fails without the `hostWidgetId ??` host-key read and passes with it. Local typecheck, lint, format, and the affected unit tests are green. Leaving mergeable for @christian-byrne to merge.

**Comment by christian-byrne** [comment]
@coderabbitai review

The Major test-hygiene item has been addressed: the mutable `hostLookup` wrapper and the always-on module mock are replaced with `vi.hoisted()` `vi.fn()` mocks plus per-test `mockImplementationOnce`/`mockReturnValueOnce` in both `useGLSLUniforms.test.ts` and `useProcessedWidgets.test.ts`. Please re-review the current head.

File: browser_tests/tests/vueNodes/glslPreview.spec.ts [hunk 5]
```diff
@@ -587,4 +588,66 @@ test.describe('GLSL Shader Preview', { tag: ['@vue-nodes', '@node'] }, () => {
...
+
+    test.beforeEach(async ({ comfyPage }) => {
+      await comfyPage.workflow.loadWorkflow(
+        'nodes/glsl_shader_subgraph_with_float'
+      )
+      await comfyPage.vueNodes.waitForNodes(1)
+    })
+
+    // Regression guard for #13875: a promoted proxy widget's live value lives at
+    // the host widget key, but the uniform source is discovered from the interior
```
Comment (DrJKL) [inline browser_tests/tests/vueNodes/glslPreview.spec.ts]: We should fix that...

File: src/renderer/extensions/vueNodes/composables/useProcessedWidgets.ts (hunk 3 above)
Comment (DrJKL) [reply src/renderer/extensions/vueNodes/composables/useProcessedWidgets.ts:195]: We should not be depending on the source node. The widget should stand alone.

File: src/renderer/glsl/useGLSLUniforms.promotedRead.test.ts [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+ * still discovered from the interior source node. `useGLSLUniforms` must read
+ * the host key so edits to the promoted curve/range proxy reach the renderer.
+ *
+ * #13875's own tests only prove `extractUniformSources` populates
+ * `hostWidgetId`; nothing pins the *read* side (`collectValues` / `curveValues`
+ * preferring `hostWidgetId`). This test closes that seam: without the read-side
+ * fix the composable returns the stale interior value.
+ */
+
+const GRAPH_ID = 'root' as UUID
```
Comment (DrJKL) [inline src/renderer/glsl/useGLSLUniforms.promotedRead.test.ts]: This is not a UUID.

File: src/renderer/glsl/useGLSLUniforms.ts (hunk 1 above)
Comment (DrJKL) [reply src/renderer/glsl/useGLSLUniforms.ts]: The near-term goal is to block _any_ traversal into Subgraphs.

Comment (DrJKL) [inline src/renderer/glsl/useGLSLUniforms.ts]: This is the source→host reverse lookup that #13434 took out of the panel code (`SectionWidgets.vue`/`WidgetActions.vue` stopped walking links back to the interior source). A fresh `findHostInputForPromotion(subgraphNode, sourceNode.id, widget.name)` call in the render path brings it back.

Good news: the value read just below now prefers `hostWidgetId`, which is the ownership direction ADR 0009 wants. The catch is how it finds that host key. The host `SubgraphNode.inputs` already carry `_subgraphSlot` and the host `widgetId`, so a host-first walk (start from the host input, dip into the interior only to line up ordering) gets you there without keying off the source. Refs ADR 0009 "Identity and value ownership" and #13434.

File: src/renderer/extensions/vueNodes/composables/useProcessedWidgets.ts [hunk 7]
```diff
@@ -330,7 +343,12 @@ export function computeProcessedWidgets({
           }
         : undefined
 
-    const nodeLocatorId = getWidgetNodeLocatorId(nodeData, bareWidgetId)
+    const nodeLocatorId = getWidgetNodeLocatorId(
+      nodeData,
+      bareWidgetId,
+      widget.sourceExecutionId,
```
Comment (DrJKL) [inline src/renderer/extensions/vueNodes/composables/useProcessedWidgets.ts:349]: Heads up: `nodeLocatorId` feeds `favoritedWidgetsStore` identity, so sourcing it from `widget.sourceExecutionId` makes a promoted widget's favorite point at the interior node. ADR 0009's app-mode/favorites section is explicit that promoted selections shouldn't persist source node/widget identity, and #13434 kept favorites on `(nodeLocatorId, widgetName)` only as a `TODO(ADR0009)` step toward a pure host `WidgetId`. This nudges it back toward the source. Better to lean on the host-scoped `widget.widgetId` (`graphId:hostNodeId:inputName`).

**Review by DrJKL — CHANGES_REQUESTED** [review]
Really nice bug hunt, and the three-layer test setup is lovely. My hesitation isn't the code, it's the direction it pulls.

The bug is real: host edits weren't reaching the GLSL preview. But the fix leans back into the source→host coupling that ADR 0009 and #13434 are busy pulling out. ADR 0009 makes promoted widgets host-scoped (`hostNodeLocator + SubgraphInput.name`) and says outright that the interior source node isn't the value owner. Source identity is fine for diagnostics, missing-model lookup, schema projection, and migration, but not for this. #13434 already took the source→host walk out of `SectionWidgets.vue`/`WidgetActions.vue` and switched to the host `widget.widgetId`.

Two spots here go the other way. I left inline notes on both.

Credit where it's due: the read path now prefers `hostWidgetId` over the interior key, which is exactly the right ownership move. I just want it to get there without the source lookup.

One honest wrinkle: uniform ordering comes from the interior node's input names, so you can't avoid reading some interior structure. My suggestion is to flip it host-first, anchor on the host inputs (they already carry `_subgraphSlot` and the host `widgetId`) and dip into the interior only to line up ordering. Happy to hop on a call and work through it if that's easier.

**Comment by DrJKL** [comment]
We need to allow Subgraph Nodes to be a black box. If this system can't work outside of the node we should find a way to mark the widgets as non-promotable.

**Comment by jtydhr88** [comment]
Since this is last-minute found issue and block our release , I prefer to keep it as temporary solution now and refactor later

**Review by DrJKL — DISMISSED** [review]
😢

File: src/renderer/glsl/useGLSLUniforms.ts (hunk 1 above)
Comment (christian-byrne) [reply src/renderer/glsl/useGLSLUniforms.ts]: Adjusted in 168db238c. Swapped the per-source `findHostInputForPromotion(subgraphNode, sourceNode.id, widget.name)` reverse lookup for a new `createPromotedHostWidgetIdLookup(subgraphNode)` that anchors on the host inputs exactly as you suggested: it iterates `subgraphNode.inputs`, takes each host `widgetId`, and walks host -> interior via `_subgraphSlot` (`resolvePromotionSource`) only to key the map by the source it projects. The render path then does an O(1) lookup instead of an O(n) reverse walk per source, and never keys off the source to *find* the host — the host input is the anchor. The value read still prefers `hostWidgetId` (the ownership direction ADR 0009 wants), so behavior is unchanged and the read-side red/green guard stays green.

Comment (christian-byrne) [reply src/renderer/glsl/useGLSLUniforms.ts]: See 168db238c: host-key resolution is now host-anchored (`createPromotedHostWidgetIdLookup`) instead of a source-keyed reverse lookup. The remaining interior read is the pre-existing GLSL uniform *discovery* (walking the shader node inputs to order uniforms), which — as you noted — is unavoidable since uniform ordering comes from the interior input names. Fully sealing subgraphs off from GLSL uniform discovery is a larger follow-up I am happy to take separately.

File: src/renderer/extensions/vueNodes/composables/useProcessedWidgets.ts (hunk 3 above)
Comment (christian-byrne) [reply src/renderer/extensions/vueNodes/composables/useProcessedWidgets.ts:195]: The locator here is used solely to look up the widget histogram from `nodeOutputStore.nodeOutputs[locatorId]` (see `WidgetCurve.vue` / `WidgetRange.vue`). That store is keyed by the node that actually executed and produced the output — for a promoted widget that is the interior source node; there is no host-keyed histogram entry to read. ADR 0009 (Identity and value ownership) explicitly keeps source identity for "diagnostics, missing-model lookup, ... only", and histogram output is exactly that class of execution-derived diagnostic, distinct from *value* ownership (which this PR correctly moved to the host key). It also reuses the existing `widget.sourceExecutionId` seam already used for promoted-widget execution errors (`hasWidgetError`, `createWidgetUpdateHandler`), so this is not new source coupling. Re-keying histogram output to the host boundary would be a `nodeOutputStore`/execution-contract change out of scope for this fix.

File: src/renderer/extensions/vueNodes/composables/useProcessedWidgets.ts (hunk 7 above)
Comment (christian-byrne) [reply src/renderer/extensions/vueNodes/composables/useProcessedWidgets.ts:349]: Verified this does not reach favorites. The `nodeLocatorId` produced by `getWidgetNodeLocatorId` lands on `SimplifiedWidget.nodeLocatorId`, and the only readers of that field are the two histogram computeds in `WidgetCurve.vue` and `WidgetRange.vue` (grep of `.nodeLocatorId` across `src`). Favorites identity is derived independently from the `LGraphNode` (`favoriteNode` in `WidgetActions.vue` / `WidgetItem.vue` -> `favoritedWidgetsStore` -> `workflowStore.nodeToNodeLocatorId(node)`), which for a promoted widget is already the host `SubgraphNode`. So this PR does not push a promoted widget favorite toward the interior node — the source-derived locator is scoped to histogram output lookup. Agreed the two identities should never be conflated; happy to add a guard/test if you want it pinned.

File: src/renderer/glsl/useGLSLUniforms.promotedRead.test.ts (hunk 6 above)
Comment (christian-byrne) [reply src/renderer/glsl/useGLSLUniforms.promotedRead.test.ts]: Fixed in c34d986a7 — `GRAPH_ID` is now a valid UUID (`3f8e2a1c-0b47-4d96-8a5e-1c2d3e4f5a6b`) instead of the `"root"` literal, and the `as UUID` cast is dropped.

File: browser_tests/tests/vueNodes/glslPreview.spec.ts (hunk 5 above)
Comment (christian-byrne) [reply browser_tests/tests/vueNodes/glslPreview.spec.ts]: Trimmed the comment in c34d986a7 to describe only the observed behavior the test guards (a promoted host edit must reach the GLSL preview), dropping the internal source/host coupling description. The e2e itself is black-box, so the host-first refactor in 168db238c does not change what it asserts. If "we should fix that" was pointing at the source->host coupling itself rather than the comment, that is addressed at the code level by the host-anchored resolver in 168db238c.

File: src/renderer/glsl/useGLSLUniforms.ts (hunk 1 above)
Comment (christian-byrne) [reply src/renderer/glsl/useGLSLUniforms.ts]: Chased both edge cases you raised on `createPromotedHostWidgetIdLookup` (168db238c). Both are already safe — no production change needed — and I pinned each with a real-fixture test in `promotionUtils.test.ts` (50e9458fb):

**1. "Would it still break if there were no interior node?"** — No. `resolvePromotionSource` is fully guarded: if the host input's `_subgraphSlot` link doesn't resolve to an interior node (removed node, or a dangling subgraph input with empty `linkIds`), it returns `undefined`, the host input is skipped, and the map simply has no entry — no throw. The lookup then returns `undefined`, and the read path falls back through `hostWidgetId ?? widgetId(gId,nId,name)` → `directValue()` → default, so the preview degrades gracefully rather than crashing. Note too that `extractUniformSources` never even reaches resolution without a live interior source node (`subgraph.getNodeById(link.origin_id)` bails first), so the "no interior node" uniform is dropped, not mis-read. Pinned by *"returns undefined (does not throw) when the interior source node is gone"* (promote → `subgraph.remove(interiorNode)` → assert no-throw + `undefined`).

**2. "Two GLSL nodes in one subgraph, each promoting a different widget — is that a thing, and does it resolve right?"** — Yes it's a supported config, and it resolves correctly with no cross-talk. The map is keyed on `(sourceNodeId \u0000 sourceWidgetName)`, and interior node ids are unique within the subgraph, so each host `widgetId` is keyed independently. Pinned by *"resolves each host widgetId independently for two sources promoting different widgets"*: two source nodes (`alpha`/`beta`) each resolve to their own distinct host id, and the cross pairs (`nodeA`+`beta`, `nodeB`+`alpha`) both return `undefined`.

Also re-confirmed the read-side red/green guard (`useGLSLUniforms.promotedRead.test.ts`) is genuinely load-bearing: dropping the two `hostWidgetId ??` reads turns it red (2 failed), restoring turns it green. Gates green (vitest/oxlint/eslint/oxfmt/typecheck); the only knip failures are the pre-existing `apps/website/*` skew, unrelated to this change.

**Comment by jtydhr88** [comment]
tested locally, works well

**Review by christian-byrne — APPROVED** [review]
